### PR TITLE
Read json customization to exclude text customizations

### PR DESCRIPTION
**Pull request by yiqunw700**
- Move design spec interfaces into a single module which is now shared by localization plugin and keywords plugin.
- Visit the reaction chain to find all the text node customizations.

### CODE REVIEW

File: support-figma/extended-layout-plugin/src/design-spec-module.ts [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+      }
+    }
+  }
+
+  var reactions = null;
+  let maybeReactions = node as ReactionMixin;
+  if (maybeReactions.reactions) {
+    reactions = maybeReactions.reactions;
+  }
+  if (reactions != null && reactions.length > 0) {
```
Comment (rylin8) [inline support-figma/extended-layout-plugin/src/design-spec-module.ts:203]: Can you add some comments to describe what is going on here, and in the handleReactionNodAsync function above? Are you following all reaction types (navigate, change, overlay, etc) and adding their destination nodes to the list of nodes to look for text customizations?

File: support-figma/extended-layout-plugin/src/design-spec-module.ts [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+    for (const customization of insideComponentDefn.customizations) {
+      // We should also look for nodes that start with "#" but don't match a customization
+      // or a top-level component name.
+      if (node.name != customization.node) continue;
+      if (
+        customization.kind == DesignCustomizationKind.Text ||
+        customization.kind == DesignCustomizationKind.TextFunction ||
+        customization.kind == DesignCustomizationKind.ComponentReplacement
+      ) {
+        if (node.type == "TEXT") {
```
Comment (rylin8) [inline support-figma/extended-layout-plugin/src/design-spec-module.ts:266]: Why not TextFunction?

File: support-figma/extended-layout-plugin/src/design-spec-module.ts [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+  } else if (insideComponentDefn) {
+    // Check to see if our name is one of the customizations, and if so, save the node if it is a text node.
+    for (const customization of insideComponentDefn.customizations) {
+      // We should also look for nodes that start with "#" but don't match a customization
+      // or a top-level component name.
+      if (node.name != customization.node) continue;
+      if (
+        customization.kind == DesignCustomizationKind.Text ||
+        customization.kind == DesignCustomizationKind.TextFunction ||
+        customization.kind == DesignCustomizationKind.ComponentReplacement
```
Comment (rylin8) [inline support-figma/extended-layout-plugin/src/design-spec-module.ts:264]: Why ComponentReplacement?

File: support-figma/extended-layout-plugin/src/ui.html [hunk 4]
```diff
@@ -255,36 +255,46 @@
...
+        <label for="uploadStringsXmlOption">Upload an existing gen_strings.xml that is generated by this plugin?</label>
+      </div>
+      <div id="stringInputPanel" style="margin-left: 18px;">
+        <div class="container" style="margin-top: 8px; margin-bottom: 8px;">
+          <button class="styled-button">
+            <label for="stringsXmlInputFile" style="width: inherit; height: inherit; line-height: 36px;">Browse</label>
+            <input type="file" id="stringsXmlInputFile" name="strings_xml_input_file" accept=".xml" class="hidden" />
+          </button>
+        </div>
+        <div id="inputStringsTableContainer" class="table-container" style="display: none; max-height: 320px;">
```
Comment (rylin8) [inline support-figma/extended-layout-plugin/src/ui.html:284]: Is this the embedded table inside the UI? It looks a little odd to me to have a table with horizontal and vertical scrollbars inside the main UI, which also has a vertical scrollbar. Is it possible to not have the inner table and make the main UI resizeable?

File: support-figma/extended-layout-plugin/src/design-spec-module.ts [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+  TextStyle = "TextStyle",
+  Image = "Image",
+  ImageWithContext = "ImageWithContext",
+  Visibility = "Visibility",
+  VariantProperty = "VariantProperty",
+  ComponentReplacement = "ComponentReplacement",
+  ContentReplacement = "ContentReplacement",
+  Modifier = "Modifier",
+  // Not covering "placeholder"; we should remove it.
+}
```
[the remark this answers is not in the document]
Comment (yiqunw700) [reply support-figma/extended-layout-plugin/src/design-spec-module.ts:32]: The snake case doesn't work for the clippy check of mismatched type. The mismatch warnings now show up when I use the camel case here.

File: support-figma/extended-layout-plugin/src/design-spec-module.ts (hunk 2 above)
Comment (yiqunw700) [reply support-figma/extended-layout-plugin/src/design-spec-module.ts:266]: This is the figma node type, it doesn't have a TextFunction type..

File: support-figma/extended-layout-plugin/src/design-spec-module.ts (hunk 3 above)
Comment (yiqunw700) [reply support-figma/extended-layout-plugin/src/design-spec-module.ts:264]: When I test with Media, there is a customization for the search field which is a ComponentReplacement. See here: https://github.com/google/automotive-design-compose/blob/031ee2816bc5cece5ab721176b15e9436dcfdcc6/reference-apps/aaos-unbundled/mediacompose/src/main/java/com/android/designcompose/reference/mediacompose/Modules.kt#L217

File: support-figma/extended-layout-plugin/src/localization-module.ts [hunk 6]
```diff
@@ -356,3 +381,41 @@ function getSavedExtras(node: TextNode): StringResourceExtras {
...
+    let customizedNodeArray: BaseNode[] = [];
+    await DesignSpecs.findTextCustomizationNodesAsync(
+      idToDesignComponentSpecMap,
+      figma.root,
+      customizedNodeArray
+    );
+    return customizedNodeArray;
+  } else {
+    figma.notify(
+      'Keywords document not found or corrupted. Run the "Check keyword definitions" plugin to upload a new keyword document.'
```
[the remark this answers is not in the document]
Comment (yiqunw700) [reply support-figma/extended-layout-plugin/src/localization-module.ts:417]: Done.

File: support-figma/extended-layout-plugin/src/ui.html (hunk 4 above)
Comment (yiqunw700) [reply support-figma/extended-layout-plugin/src/ui.html:284]: I removed height limit of the table so it won't have a vertical scrollbar. But I don't know how to make the main UI resizable that it seems figma's api ShowUIOptions doesn't have a way to make it resizable.

File: support-figma/extended-layout-plugin/src/design-spec-module.ts (hunk 1 above)
Comment (yiqunw700) [reply support-figma/extended-layout-plugin/src/design-spec-module.ts:203]: Done.

File: support-figma/extended-layout-plugin/src/ui.html [hunk 7]
```diff
@@ -255,36 +255,46 @@
...
+        </div>
+      </div>
     </div>
+  
+    <label id="showOutputsButton"
+      style="margin-top: 16px; position: fixed; bottom: 20px; right: 20px; text-decoration: underline; color: blue; line-height: 36px;"
+      onclick="showOutputStrings()"> Next >> </label>
```
[the remark this answers is not in the document]
Comment (yiqunw700) [reply support-figma/extended-layout-plugin/src/ui.html]: I broke it into two pages. The first page only has the exclude options and the second page remain the same where it has a "Skip" button and can change to "Next" when it uploads a string xml file.

**Comment by rylin8** [comment]
Not really related to this change, but the "Submit" button is a little confusing as to whether it applies to the "Char Limit" and "Description", or just "Description".

**Review by rylin8 — APPROVED** [review]

**Pull request merged**